modules/conversantAnalyticsAdapter.js

```diff
@@ -0,0 +1,470 @@
+      'auctionId': auctionId,
+      'timeReceived': Date.now() // For cache cleaning
+    };
+  }
+
+  sendData(bidWonPayload);
+}
+
+/**
+ * RENDER FAILED occurs after AUCTION END and BID WON, the payload does not have all the data we need so we use
+ * adIdLookup to pull data from a BID WON event to populate our payload
+ * @param args = {
+ *  reason: <value>
+ *  message: <value>
+ *  adId: <value> --optional
+ *  bid: {object?} --optional: unsure what this looks like but guessing it is {bidder: <value>, params: {object}}
+ *    }
+ */
+function onAdRenderFailed(args) {
+  const adId = args.adId;
+  // Make sure we have all the data we need, adId is optional so it's not guaranteed, without that we can't match it up
+  // to our adIdLookup data.
+  if (typeof adId === 'undefined' || typeof adIdLookup[adId] === 'undefined') {
+    utils.logError(LOG_PREFIX + "onAdRenderFailed(): Unable to process RENDER FAILED because adId is missing or doesn't match a record in our cache");
+    return; // Either no adId to match against a bidWon event, or no data saved from a bidWon event that matches the adId
+  }
+  const adIdObj = adIdLookup[adId];
+  const adUnitCode = adIdObj['adUnitCode'];
+  const bidderCode = adIdObj['bidderCode'];
+  const auctionId = adIdObj['auctionId'];
+  delete adIdLookup[adId]; // cleanup our cache
+
+  if (typeof bidderCode === 'undefined' || typeof adUnitCode === 'undefined') {
+    return;
+  }
+
+  const renderFailedPayload = createPayload('render_failed', auctionId);
+  const adUnitPayload = createAdUnit();
+  adUnitPayload.bids[bidderCode] = initializeBidDefaults();
+  adUnitPayload.bids[bidderCode].eventCodes.push(CNVR_RENDER_FAILED);
+  adUnitPayload.bids[bidderCode].message = 'REASON: ' + args.reason + '. MESSAGE: ' + args.message;
+  renderFailedPayload.adUnits[adUnitCode] = adUnitPayload;
+  sendData(renderFailedPayload);
+}
+
+/**
+ * AUCTION END contains bid and no bid info and all of the auction info we need. This sends the bulk of the information
+ * about the auction back to the servers.  It will also check the timeoutCache for any matching bids, if any are found
+ * then they will be removed from the cache and send back with this payload.
+ * @param args AUCTION END payload, fairly large data structure, main objects are 'adUnits[]', 'bidderRequests[]',
+ * 'noBids[]', 'bidsReceived[]'... 'winningBids[]' seems to be always blank.
+ */
+function onAuctionEnd(args) {
+  const auctionId = args.auctionId;
+  if (typeof auctionId === 'undefined') {
+    utils.logError(LOG_PREFIX + 'onAuctionEnd(): No auctionId in args supplied so unable to process event.');
+    return;
+  }
+
+  const auctionEndPayload = createPayload('auction_end', auctionId);
+  // Get bid request information from adUnits
+  if (!Array.isArray(args.adUnits)) {
+    utils.logError(LOG_PREFIX + 'onAuctionEnd(): adUnits not defined in arguments.');
+    return;
+  }
+
+  args.adUnits.forEach(adUnit => {
+    const cnvrAdUnit = createAdUnit();
```

Comment:
Calling our stuff adUnit, too,  is pretty confusing and probably misleading at times.  Maybe we should call it something else like unitInfo, unitData, unitPayload etc.  So createAdUnit() should be createUnitData().  Thoughts?

Reply:
I see your point, I'm a bit too close to the code to be objective.  My thought is with this very loose data structure we want to have similar words so I know what that what I'm creating is essentially the same data structure as what I was given.  Calling them different names feels like they are fundamentally different and makes the code more difficult to follow.  Just my 2 cents though.

Reply:
That's fine.